### PR TITLE
Use Groovy 2.4.15

### PR DESCRIPTION
**Pull request by lptr**
Let's try this again...

### CODE REVIEW

**Comment by lptr** [comment]
Branch build: https://builds.gradle.org/viewQueued.html?itemId=14331006

**Comment by lptr** [comment]
Branch build: https://builds.gradle.org/viewQueued.html?itemId=14354747

**Comment by lptr** [comment]
@wolfs can you please take a look? I ended up adding some comments to make it easier to upgrade Groovy in the future. I've also bumped the Groovy versions used in samples, but only to avoid promoting a known buggy compiler version. We don't need to bump that version all the time.

File: buildSrc/build.gradle.kts [hunk 1]
```diff
@@ -154,6 +154,7 @@ fun Project.applyGroovyProjectConventions() {
         testCompile("cglib:cglib:3.2.6")
         testCompile("org.objenesis:objenesis:2.4")
         constraints {
+            // This must match the Groovy version bundled with the Gradle wrapper used to build the project
             compile("org.codehaus.groovy:groovy-all:2.4.12")
```
Comment (wolfs) [inline buildSrc/build.gradle.kts]: Can't we somehow get the Groovy version for the Gradle wrapper?

File: subprojects/base-services/src/test/groovy/org/gradle/internal/service/ServiceLocatorTest.groovy [hunk 2]
```diff
@@ -115,7 +115,7 @@ org.gradle.ImplClass2""")
 
         then:
         RuntimeException e = thrown()
-        e.message == "Could not determine implementation class for service 'java.lang.String' specified in resource '${serviceFile}'."
+        e.message == "Could not determine implementation class for service 'java.lang.String' specified in resource '" + serviceFile + "'."
```
Comment (wolfs) [inline subprojects/base-services/src/test/groovy/org/gradle/internal/service/ServiceLocatorTest.groovy]: Why not use GStrings any more? Did something change in Groovy about equality?

File: subprojects/docs/src/samples/groovy/crossCompilation/build.gradle [hunk 3]
```diff
@@ -1,3 +1,5 @@
+import org.apache.tools.ant.taskdefs.Javadoc
```
Comment (wolfs) [inline subprojects/docs/src/samples/groovy/crossCompilation/build.gradle]: I don't think we should have this import.

File: buildSrc/build.gradle.kts (hunk 1 above)
Comment (lptr) [reply buildSrc/build.gradle.kts]: Yeah, I just got that idea, too :)

File: subprojects/base-services/src/test/groovy/org/gradle/internal/service/ServiceLocatorTest.groovy (hunk 2 above)
Comment (lptr) [reply subprojects/base-services/src/test/groovy/org/gradle/internal/service/ServiceLocatorTest.groovy]: Yeah, a null value doesn't seem to evaluate to `null` but to an empty string now.

**Comment by lptr** [comment]
Experimental build: https://builds.gradle.org/viewQueued.html?itemId=14376912

**Comment by lptr** [comment]
Branch build: https://builds.gradle.org/viewQueued.html?itemId=14381041

**Comment by oehme** [comment]
There already is a regression on master in that test, did you take that into account?

**Comment by lptr** [comment]
I see. So it seems we've gone from 35/48 ms to 64 ms overhead, which is even less than before. I'll merge the PR as soon as we have a green `master`.

**Pull request merged**